File: src/main/java/calebzhou/rdi/microservice/component/RdiResponseAdvice.java
package calebzhou.rdi.microservice.component;

import calebzhou.rdi.microservice.constant.ResponseCode;
import lombok.SneakyThrows;
import org.springframework.core.MethodParameter;
import org.springframework.http.MediaType;
import org.springframework.http.converter.HttpMessageConverter;
import org.springframework.http.server.ServerHttpRequest;
import org.springframework.http.server.ServerHttpResponse;
import org.springframework.web.bind.annotation.RestControllerAdvice;
import org.springframework.web.servlet.mvc.method.annotation.ResponseBodyAdvice;


@RestControllerAdvice
public class RdiResponseAdvice implements ResponseBodyAdvice<Object> {
    @Override
    public boolean supports(MethodParameter returnType, Class<? extends HttpMessageConverter<?>> converterType) {
        return true;
    }

    @SneakyThrows
    @Override
    public Object beforeBodyWrite(Object body, MethodParameter returnType, MediaType selectedContentType, Class<? extends HttpMessageConverter<?>> selectedConverterType, ServerHttpRequest request, ServerHttpResponse response) {
        if(body instanceof String st)
            return ResponseData.success(st);
        if(body instanceof ResponseData<?>)
            return body;
        if(body instanceof ResponseCode responseCode)
            return responseCode.toResultData();
        if(body instanceof Enum e)
            return ResponseData.success(e.toString());
        /*if(String.valueOf(body).contains("Not Found") && String.valueOf(body).contains("404"))
            return ResultCode.notFound.toResultData();*/
        return ResponseData.success(body);
    }
}

File: src/main/kotlin/calebzhou/rdi/microservice/component/RdiResponseAdvice.kt
package calebzhou.rdi.microservice.component

import calebzhou.rdi.microservice.constant.ResponseCode
import calebzhou.rdi.microservice.model.ResponseData
import org.springframework.core.MethodParameter
import org.springframework.http.MediaType
import org.springframework.http.converter.HttpMessageConverter
import org.springframework.http.server.ServerHttpRequest
import org.springframework.http.server.ServerHttpResponse
import org.springframework.web.bind.annotation.RestControllerAdvice
import org.springframework.web.servlet.mvc.method.annotation.ResponseBodyAdvice

/**
 * Created by calebzhou on 2022-10-05,7:24.
 */
@RestControllerAdvice
class RdiResponseAdvice : ResponseBodyAdvice<Any> {
    override fun supports(returnType: MethodParameter, converterType: Class<out HttpMessageConverter<*>>): Boolean {
        return true
    }

    override fun beforeBodyWrite(
        body: Any?,
        returnType: MethodParameter,
        selectedContentType: MediaType,
        selectedConverterType: Class<out HttpMessageConverter<*>>,
        request: ServerHttpRequest,
        response: ServerHttpResponse
    ): Any {
        return when (body) {
            is ResponseData<*> -> body
            is ResponseCode -> ResponseData(body,null)
            else -> ResponseData(ResponseCode.Success,body)
        }

    }
}